### ISSUE TITLE
Repeating column labels for groups (a la "Freeze Panes" on Excel)

### ISSUE THREAD

**Issue by gitmoneyMSBA** [rstudio/gt#194]
---

Thanks so much for the development and support of this package. It's the best table package we have on R and it renders beautifully to web pages and reports.

I'm often using it for larger tables split into two groups (>10 observations for each group), so the _n_ and _p_ are pretty close (somewhere between 10 and 15 variables). 

Excel has a **freeze panes** option that allows you to scroll down a long table/grouped table and not lose track of which variable you're tracking. 

I'm not sure if there's functionality to mimic that in grouped tables at the moment (hopefully there is and I've just missed this!), but it seems worthwhile for longer tables if there's not.

Imagine if country pops had many variables about population, and we could repeat the tab spanners/column names above both United States and Russian Federation). Example included has a wishful function called **col_name_repeat**.

```r
library(gt)

country = countrypops %>% filter(country_name %in% c("United States", "Russian Federation"),  year >= 1985)

gt_tbl <-  gt(data = country ) %>%
    group_by(country_name) %>%
  tab_spanner(
    label = "Population statistics",
    columns = vars(population, 2:10)
  ) %>%
  tab_spanner(
    label = "Political system/info",
    columns = vars(11:19)
  ) %>%
  col_name_repeat(groups=c("United States", "Russian Federation")

```

**Comment by markbneal** [comment]
Does this relate to this issue? [412](https://github.com/rstudio/gt/issues/412)
DT handles this with extensions, what it calls FixedColumns and FixedHeaders, described [here](https://rstudio.github.io/DT/extensions.html)

**Comment by sumairshah2** [comment]
Is there anything like this available now

**Comment by mbsteel** [comment]
I, too, would be interested in this feature as I need to produce long reports on several variables for the 120 counties in Kentucky and would appreciate an update if it has or is being developed. Please and thank you!

**Comment by sumairshah2** [comment]
@mbsteel I figured out one way to manually do this. I do not think it has been developed to my knowledge.

**Comment by mbsteel** [comment]
@sumairshah2 I would appreciate seeing your method, if you don't mind. Otherwise, I might need to resort to creating multiple tables, one for each page.

**Comment by sumairshah2** [comment]
[targetlogs.pdf](https://github.com/rstudio/gt/files/6473616/targetlogs.pdf)

to confirm -- you are looking to do something similar to the report I attached? Repeat column names after every row group?

**Comment by mbsteel** [comment]
@sumairshah2 Yes, that would work for me.

**Comment by sumairshah2** [comment]
Shoot me an email - sumairshah@gatech.edu

**Comment by gitmoneyMSBA** [comment]
Would you be willing to share your solution here? That’s exactly what I’m looking for in terms of a workaround.

**Comment by sumairshah2** [comment]
Yeah, let me write some example code and share it here. With my report I used above I have also started using https://gt.rstudio.com/reference/cols_hide.html to hide_columns. Makes it more aesthetic.  You can see that in this report 
![image](https://user-images.githubusercontent.com/71948524/118194977-2bee4500-b3ff-11eb-809e-79e47a798041.png)

**Comment by stevenbedrick** [comment]
I think this would be a super-useful feature, as well, and like the proposed `col_name_repeat()` syntax. @sumairshah2 , if you could post the code for your manual solution, I'd be really curious to see it!

**Comment by mbsteel** [comment]
In case it helps, I ended up using the kable package to produce repeating tables. But I do like some of the features in gt better and wish there was a way to use that package to the same effect.

**Comment by sumairshah2** [comment]
Hi @mbsteel @stevenbedrick @gitmoneyMSBA  -- sorry for the late response. Paste this into R and let me know if it makes sense. Super hacky way of doing this and after seeing the logic you may find a cleaner way. Let me know if you have any questions or run into trouble.  You should get this result below
![Rplot](https://user-images.githubusercontent.com/71948524/125866971-58fe00c8-3986-4d9f-9019-4fdc01ad07a0.png)


```
library(janitor)
library(gt)
library(dplyr)

# Just used this as a built-in R data set
B <- Lahman::Batting %>% tail(5)
# Take the colnames
C <- colnames(B)

## you can probably figure this part out better but for now, rbind colnames * length(df)  
D <- as.data.frame(rbind(C,C,C,C,C),row.names = F,make.names = F) %>% janitor::row_to_names(1,remove_row = F)

## Replace our column df with a vector of real data from the original df
D$playerID <- B$playerID

# combine colnames and real data
final_df <- rbind(D,B)


## you can hide or not hide column names. 

final_df %>% arrange(desc(yearID)) %>% gt(groupname_col ="playerID")%>% tab_header(title = md("**Advisees**"),)  %>% tab_options(heading.background.color = "#EFFBFC",stub.border.style = "dashed",stub.border.color = "#989898",stub.border.width = "1px",summary_row.border.color = "#989898",table.width = "75%",grand_summary_row.background.color = "Navy",column_labels.background.color = "black",table.font.color = "black",row_group.border.bottom.color = "black",row_group.border.bottom.width = 2, row_group.padding = 10,row_group.background.color = "#EFFBFC",stub.font.weight = "bold") %>% tab_options(column_labels.hidden = F) %>% tab_style(style = list(cell_fill(color = "Grey")),locations = cells_body(rows = yearID == "yearID"))







```

**Comment by mbsteel** [comment]
Thanks, will give it a try!

Sent from my Sprint Samsung Galaxy S10e.
Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by stevenbedrick** [comment]
Thanks, @sumairshah2 ! I had been wondering if your trick was something along these lines (adding placeholder/dummy rows to serve as the headers), and had been starting to think about how I might do something similar in my situation but your way of doing it is much cleverer than what I had been coming up with. :-) Thanks so much for sharing!

**Comment by markbneal** [comment]
Hi, I've just made some mods to @sumairshah2 code to make some of the steps a bit clearer, might be useful to others? Apologies for the base R in places! 

It seems quite possible that if you apply number formatting to the table, there might be a problem, because all columns in the final data frame end up as character. 

```
#gt repeat headings by group
#https://github.com/rstudio/gt/issues/194#issuecomment-881056646

library(janitor)
library(gt)
library(dplyr)

# Just used this as a built-in R data set
my_data <- Lahman::Batting %>% tail(5)

# TO DO: make groupname as variable, then refer to throughout with {{}}?
# group_name

# 
groups <- my_data %>% 
  select(playerID) %>% 
  distinct()

# Take a count of groups
group_number <- my_data %>% 
  select(playerID) %>% 
  n_distinct()

## Make dummy data frame of headers, option 1
# Take the colnames
my_colnames <- colnames(my_data)

dummy_header_data <- data.frame(t(my_colnames)) %>% 
  janitor::row_to_names(1,remove_row = F)

#make a row for each group_number 
dummy_header_data[1:group_number,] <- dummy_header_data[1,]


## Make dummy data frame of headers, option 2

my_data_empty <- my_data[FALSE,]

my_data_empty[] <- lapply(my_data_empty, as.character)

dummy_header_data[1:group_number,] <- data.frame(t(colnames(my_data_empty)))

#purrr::map_dfr(seq_len(group_number), ~my_colnames) #more tidyverse-y, but not working


## Replace our column df with a vector of real data from the original df
dummy_header_data$playerID <- groups[,1]

# combine colnames and real data
final_df <- bind_rows(dummy_header_data, my_data %>% mutate_all(as.character))


## you can hide or not hide column names. 

#basic gt table
final_df %>% arrange(desc(yearID)) %>% 
  gt(groupname_col ="playerID")

#formatted gt table
final_df %>% arrange(desc(yearID)) %>% 
  gt(groupname_col ="playerID") %>% 
  tab_header(title = md("**Advisees**"),)  %>% 
  tab_options(
              heading.background.color = "#EFFBFC", 
              stub.border.style = "dashed", 
              stub.border.color = "#989898",
              stub.border.width = "1px", 
              summary_row.border.color = "#989898",
              table.width = "75%", 
              grand_summary_row.background.color = "Navy", 
              column_labels.background.color = "black",
              table.font.color = "black", 
              row_group.border.bottom.color = "black",
              row_group.border.bottom.width = 2, 
              row_group.padding = 10,
              row_group.background.color = "#EFFBFC",
              stub.font.weight = "bold"
              ) %>% 
  tab_options(column_labels.hidden = F) %>% 
  tab_style(style = list(cell_fill(color = "Grey")),
            locations = cells_body(rows = yearID == "yearID"))
```

**Comment by stevenbedrick** [comment]
And to join the fun, here's what I came up with for making the data frame of dummy rows! It's not quite as elegant as the example above; I used `rep()` on the vector of colnames and then `matrix()`'d it, rather than the clever multi-row-assignment syntax that @markbneal  used:

```
# first figure out how many groups to add, via n_distinct() or however one wishes
num.groups <- my_data %>% select(grouping.column) %>% n_distinct()

# make a matrix of the column names, one per group
dummy.rows.mat <- matrix(rep(column.names, num.groups), nrow=num.groups, byrow=TRUE)

# turn into a data frame for rbinding with the original data 
dummy.rows.df <- as.data.frame(dummy.rows.mat, row.names=F, make.names=F) %>% janitor::row_to_names(1, remove_row = F)
```

For my particular use case, I only wanted to repeat certain column headers so I also have a step where I `mutate()` certain columns in `dummy.rows.df` into NAs, and then I use `fmt_missing()` to deal with those later in the table creation.

On an unrelated note, I would like to thank @rich-iannone and the rest of the `gt` devs for an amazing library. I'm working on a particularly gnarly table, and this is literally one of two things I've wanted to do with my table that I haven't been able to do using `gt`'s built-in functionality. And thanks to the library's flexibility, this workaround ended up being conceptually pretty straightforward, which I see as being a real credit to the library and its core design.

**Comment by ArthurAndrews** [comment]
This would be a great feature.  Similar to the `fixed_thead = TRUE` option in kable_extra.

**Comment by jonathanmburns** [comment]
I agree that having the ability to lock the header row would be extremely helpful.  I have a large table right now in columns and rows; the fact that the column headers get lost when scrolling up definitely minimizes the usability of the table to the end users.  I know that there are other options (as mentioned in this thread), but I would like to stay with the `gt` package.

**Comment by dmontecino** [comment]
This would be a great feature!

**Comment by RSA-Workspace** [comment]
The problem is related to overflow property set in the outer div and .cell-output-display. Just unset them and it works.

`mtcars %>% gt(id="two") %>%
  opt_css(
    css = "
    .cell-output-display {
      overflow-x: unset !important;
    }
    div#two {
      overflow-x: unset !important;
      overflow-y: unset !important;
    }
    #two .gt_col_heading {
      position: sticky !important;
      top: 0 !important;
    }
    "
  )`

**Comment by olivroy** [comment]
This is natively possible with reactable (i.e. with `opt_interactive()` )


`reactable::colGroup(sticky = TRUE)` could be used for `columnGroups`
